src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/CSharpOnTypeFormattingPass.cs

```diff
@@ -0,0 +1,130 @@
+            ILoggerFactory loggerFactory)
+            : base(documentMappingService, filePathNormalizer, server)
+        {
+            if (loggerFactory is null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<CSharpOnTypeFormattingPass>();
+        }
+
+        public async override Task<FormattingResult> ExecuteAsync(FormattingContext context, FormattingResult result)
+        {
+            if (!context.IsFormatOnType || result.Kind != RazorLanguageKind.CSharp)
+            {
+                // We don't want to handle regular formatting or non-C# on type formatting here.
+                return result;
+            }
+
+            // Normalize and re-map the C# edits.
+            var codeDocument = context.CodeDocument;
+            var csharpText = SourceText.From(codeDocument.GetCSharpDocument().GeneratedCode);
+            var normalizedEdits = NormalizeTextEdits(csharpText, result.Edits);
+            var mappedEdits = RemapTextEdits(codeDocument, normalizedEdits, result.Kind);
+
+            // Find the lines that were affected by these edits.
+            var originalText = codeDocument.GetSourceText();
+            var changes = mappedEdits.Select(e => e.AsTextChange(originalText));
+            var changedText = originalText.WithChanges(changes);
+            TrackEncompassingChange(originalText, changedText, out var spanBeforeChange, out var spanAfterChange);
+            var rangeBeforeEdit = spanBeforeChange.AsRange(originalText);
+            var rangeAfterEdit = spanAfterChange.AsRange(changedText);
+
+            // Create a new formatting context for the changed razor document.
+            var changedContext = await context.WithTextAsync(changedText);
+
+            // Now, for each affected line in the edited version of the document, remove x amount of spaces
+            // at the front to account for extra indentation applied by the C# formatter.
+            // This should be based on context.
+            // For instance, lines inside @code/@functions block should be reduced one level
+            // and lines inside @{} should be reduced by two levels.
+            var indentationChanges = AdjustCSharpIndentation(changedContext, (int)rangeAfterEdit.Start.Line, (int)rangeAfterEdit.End.Line);
+
+            // Apply the edits that remove indentation.
+            changedText = changedText.WithChanges(indentationChanges);
+            changedContext = await changedContext.WithTextAsync(changedText);
+
+            // We make an optimistic attempt at fixing corner cases.
+            changedText = CleanupDocument(changedContext, rangeAfterEdit);
+
+            // Now that we have made all the necessary changes to the document. Let's diff the original vs final version and return the diff.
+            var finalChanges = SourceTextDiffer.GetMinimalTextChanges(originalText, changedText, lineDiffOnly: false);
```

Comment:
Possibly for huge files. But we can't afford to replace whole lines in onTypeFormatting because it can potentially break the typing experience by moving the cursor to unexpected locations. We'll optimize based on user feedback.